src/main/java/com/derongan/minecraft/deeperworld/listeners/MovementListener.kt

```diff
@@ -81,10 +101,100 @@ object MovementListener : Listener {
 
     private fun teleportBetweenSections(player: Player, to: Location, oldSection: Section, newSection: Section) {
         val newLoc = to.getCorrespondingLocation(oldSection, newSection) ?: return
-        val fallDistance = player.fallDistance
-        val oldVelocity = player.velocity
-        player.teleport(newLoc)
-        player.fallDistance = fallDistance
-        player.velocity = oldVelocity
+
+        val oldLeashedEntities = player.getLeashedEntities()
+        oldLeashedEntities?.forEach {
+            if (it.customName.isNullOrEmpty()) {
+                it.customName = " "
+            }
+        }
+
+        val rootVehicle = player.getVehicleRecursive()
+        if (rootVehicle != null) {
+            newLoc.yaw = player.location.yaw
+            newLoc.pitch = player.location.pitch
+
+            // Prevent teleportation of other players in the vehicle-passenger structure.
+            rootVehicle.getPlayerPassengersRecursive()?.forEach {
+                if (it != player) {
+                    it.vehicle?.removePassenger(it)
+                }
+            }
+
+            val oldFallDistance = rootVehicle.fallDistance
+            val oldVelocity = rootVehicle.velocity
+
+            val vehicleTree = VehicleTree(rootVehicle)
+
+            vehicleTree.root.applyAll {
```

Comment:
Maybe add a comment explaining we remove passengers beforehand to avoid the bug where vehicles don't get rendered.